theano/sandbox/cuda/nnet.py

```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
         return ret1 + "\n" + ret2
 
 gpu_softmax_with_bias = GpuSoftmaxWithBias()
+
+
+def hierarchical_softmax(W1, b1, W2, b2, x, n_outputs, target=None):
+    """
+    GPU-only function that returns the outputs of a two-level hierarchical
+    softmax.
+    In the two-level hierarchical softmax architecture, outputs are grouped
+    in sqrt(n_outputs) classes.
+    There are two softmax layers. The first predicts the class of the input x
+    while the second predicts the output of the input x in the predicted class.
+    More explanations can be found in the original paper
+    http://arxiv.org/abs/cs/0108006.
+    If target is specified, it will only compute the outputs of the
+    corresponding targets. Otherwise, if target is None, it will compute all
+    the outputs.
+
+    :type W1: a Theano shared variable of shape (number of features of the
+        input x, number of classes)
+    :param W1: the weight matrix of the first softmax, which maps the
+        input x to the probabilities of the classes.
+
+    :type b1: a Theano shared variable of shape (number of classes,)
+    :param b1: the bias vector of the first softmax layer.
+
+    :type W2: a Theano shared variable of shape (number of classes, number of
+        features of the input x, number of outputs per class)
+    :param W2: the weight matrix of the second softmax, which maps the input
+        x to the probabilities of the outputs.
+
+    :type b2: a Theano shared variable of shape (number of classes, number of
+        outputs per class)
+    :param b2: the bias vector of the second softmax layer.
+
+    :type x: Theano tensor variable of shape (batch_size, number of features)
+    :param x: the minibatch input of the multi-class softmax.
+
+    :type n_outputs: int
+    :param n_outputs: the number of outputs.
+
+    :type target: a Theano variable of shape either (batch_size,) or
+        (batch_size, 1)
+    :param target: Contains the indices of the targets for the minibatch
+        input x. For each input, the function computes the output for its
+        corresponding target. If target is None, then all the outputs are
+        computed for each input.
+
+    :rtype: Theano tensor variable of the following shape:
+        - (batch_size, 1) if target is not None
+        - (batch_size, n_out) if target is None
+    :returns: the minibatch output of the 2-layer hierarchical softmax.
+
+    :note: If n_outputs is not a square number, it will still work but the
+        user has to specify the following shapes of its weight/bias matrices:
+        - number of outputs per class: n_outputs_per_class = ceil(sqrt(
+          n_outputs)
+        - number of classes: n_classes = ceil(n_outputs / n_outputs_per_class)
+
+    """
+
+    if theano.config.device != 'gpu':
```

Comment:
It seems better to test [this way](https://github.com/Theano/Theano/blob/master/theano/sandbox/cuda/tests/test_nnet.py#L12)